aptos-move/framework/aptos-framework/sources/Coin.move

```diff
@@ -307,6 +307,20 @@ module AptosFramework::Coin {
         move_to(account, coin_store);
     }
 
+    /// "Route" `amount` of coins of `CoinType` from `from` to `to`.
+    /// Like `transfer()` but for standalone `Coin` instances not kept
+    /// in a `CoinStore` (alternatively, like `merge()`, but for an
+    /// in-place extraction from `source_coin`). Automatically aborts
+    /// for insufficient `value` held in `from`, per integer underflow.
+    public fun route<CoinType>(
+        from: &mut Coin<CoinType>,
+        to: &mut Coin<CoinType>,
+        amount: u64
+    ) {
+        from.value = from.value - amount;
+        to.value = to.value + amount;
+    }
+
```

Comment:
If we really need this, it would probably be better to use extract and merge, notice the  semantics are different.

Reply:
@davidiw as you suggested, the proposed routing functionality could indeed be achieved via:

```move
let intermediate = extract<CoinType>(from, amount);
merge<CoinType>(to, intermediate);
```

or, more succinctly:

```move
merge<CoinType>(to, extract<CoinType>(from, amount));
```

but alas this approach introduces the overhead steps of

1. [packing](https://github.com/aptos-labs/aptos-core/blob/8c5882081f42ee5cfed19dd32de256790f250419/aptos-move/framework/aptos-framework/sources/Coin.move#L210)
2. [returning](https://github.com/aptos-labs/aptos-core/blob/8c5882081f42ee5cfed19dd32de256790f250419/aptos-move/framework/aptos-framework/sources/Coin.move#L207),
3. [passing](https://github.com/aptos-labs/aptos-core/blob/8c5882081f42ee5cfed19dd32de256790f250419/aptos-move/framework/aptos-framework/sources/Coin.move#L256),
4. then [unpacking](https://github.com/aptos-labs/aptos-core/blob/8c5882081f42ee5cfed19dd32de256790f250419/aptos-move/framework/aptos-framework/sources/Coin.move#L258) an intermediate `Coin`,

all of which are avoided in the proposed `route()` function.
At present it is unclear how much compute/gas/time would be consumed via such redundancies in a single execution step, but of course it would quickly add up should such routing be performed hundreds or thousands of times per second (notably, like when routing funds between users in an order book application, for instance from one [`Econia::User::OC`](https://github.com/econia-labs/econia/blob/1235179bfc7fc98a4dc71ee18ce0989147e696ea/src/move/econia/sources/User.move#L77) order collateral container to another, the use case that originally modified this pull request).

Your thoughts?

Reply:
That's a fair question, I would err on the side of not optimizing just yet. I'm hopeful we can get a decent evaluator pretty soon -- that would take into consideration these things. Then we should review the entirety of the framework and simplify / optimize where appropriate. I suspect we'll find a lot of "interesting" things out.

Reply:
Yes an evaluator would certainly make that part of the design process easier. As for the functionality itself, whether optimized or unoptimized, @the-fool indicated in a separate Discord thread that he too has a use case for the proposed `route()` paradigm, and more broadly this kind of feature would probably serve as a primitive for applications that manage `Coin` types outside of a `CoinStore`.

As for semantics, however, perhaps one barrier with the proposed implementation is the semantics of arguments? At present it accepts `from` and `to`, analogous to `transfer()`, which similarly moves value from one `Coin` to another, however unlike `transfer()`, `route()` operates on standalone `Coin` types like `extract()` and `merge()`. Hence perhaps a more function signature would be as follows?

```move
    public fun route<CoinType>(
         source_coin: &mut Coin<CoinType>,
         dst_coin: &mut Coin<CoinType>,
         amount: u64
     ) {
         source_coin.value = source_coin.value - amount;
         dst_coin.value = dst_coin.value + amount;
     }
```

Here, it accepts `source_coin` and `dst_coin` as parameters in the style of `merge()`.